### ISSUE TITLE
Training seems to hang sometimes

### ISSUE THREAD

**Issue by wvalcke** [ultralytics/ultralytics#673]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Sometimes training just hangs, there is no error at all, it's just stalled in the mid of an epoch.
I had this already twice, and when doing CTRL-C it shows this traceback.
If i have it again i'll update this report, hopefully the traceback will be the same.
To me it looks it's stuck somewhere waiting in the dataloader.
```
Traceback (most recent call last):
  File "./yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/cfg/__init__.py", line 237, in entrypoint
    func(cfg)
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/v8/detect/train.py", line 205, in train
    model.train(**vars(cfg))
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/engine/model.py", line 201, in train
    self.trainer.train()
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/engine/trainer.py", line 183, in train
    self._do_train(int(os.getenv("RANK", -1)), world_size)
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/engine/trainer.py", line 284, in _do_train
    for i, batch in pbar:
  File "/home/wim/.virtualenvs/pytorch1.12.1/lib/python3.8/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/home/wim/.virtualenvs/pytorch1.12.1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 681, in __next__
    data = self._next_data()
  File "/home/wim/.virtualenvs/pytorch1.12.1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1359, in _next_data
    idx, data = self._get_data()
  File "/home/wim/.virtualenvs/pytorch1.12.1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1315, in _get_data
    success, data = self._try_get_data()
  File "/home/wim/.virtualenvs/pytorch1.12.1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1163, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/usr/lib/python3.8/queue.py", line 179, in get
    self.not_empty.wait(remaining)
  File "/usr/lib/python3.8/threading.py", line 306, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by rusamentiaga** [comment]
I had the same problem, but only when using cache=ram. There could be a thread synchronization issue when using the memory cache.

**Comment by wvalcke** [comment]
Good remark, i also use cache=ram option.

**Comment by leoncch** [comment]
I faced the same problem, but I dont use cache=ram option.
Here is my training command: yolo detect train data=data/safety_vest_dataset_V2.1/vest2.1.yaml model=weights/yolov8x.pt epochs=300 imgsz=640 batch=16 device=0 name=v8_vest2.1

I remember that I used the same command but with epochs=100 and finished my training successfully but I faced the same problem below after training, so I update my torch version to 1.12.0+cu116, maybe the hanging was caused by this???
[#6962](https://github.com/ultralytics/yolov5/issues/6962)

**Comment by wvalcke** [comment]
I upgraded my system to Linux mint 21.1, using CUDA 11.7 and python 3.10.6 with pytorch 1.13.1, using latest git main head.
```
commit 3633d4c06b806a971f444919c389ac6e889cdb23 (HEAD -> main, origin/main, origin/HEAD)
Author: Glenn Jocher <glenn.jocher@ultralytics.com>
Date:   Tue Feb 7 02:37:13 2023 +0400

    TAL `min_memory` argument, precommit, Docker fixes (#836)
    
    Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
    Co-authored-by: Jaap van de Loosdrecht <jaap@vdlmv.nl>
```
The problem remains, and is even worse now. It hangs regularly in the same method.
The Dataloader hangs waiting on something related to threading.
```
Traceback (most recent call last):
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/./yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/cfg/__init__.py", line 266, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/engine/model.py", line 214, in train
    self.trainer.train()
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/engine/trainer.py", line 185, in train
    self._do_train(int(os.getenv("RANK", -1)), world_size)
  File "/mnt/DV/wim/BEP/Software/DeepLearningStudy/ultralytics/ultralytics/yolo/engine/trainer.py", line 286, in _do_train
    for i, batch in pbar:
  File "/home/wim/.virtualenvs/pytorch1.13.1/lib/python3.10/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/home/wim/.virtualenvs/pytorch1.13.1/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 628, in __next__
    data = self._next_data()
  File "/home/wim/.virtualenvs/pytorch1.13.1/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1316, in _next_data
    idx, data = self._get_data()
  File "/home/wim/.virtualenvs/pytorch1.13.1/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1272, in _get_data
    success, data = self._try_get_data()
  File "/home/wim/.virtualenvs/pytorch1.13.1/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1120, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/usr/lib/python3.10/queue.py", line 180, in get
    self.not_empty.wait(remaining)
  File "/usr/lib/python3.10/threading.py", line 324, in wait
    gotit = waiter.acquire(True, timeout)
```
waiter.acquire -> thats where it get stuck.

Training always takes place with 1 single GPU. 
Sometimes it hangs after 30 epochs, sometimes after 500, it's really time dependent and is definitely a race condition.

**Comment by wvalcke** [comment]
More info, looks like a pytorch issue....

https://discuss.pytorch.org/t/dataloader-freezes/159155

**Comment by Raltrwx** [comment]
I think I'm having the same issue where training just pauses/stops for seemingly no reason, I just re-started my training using a lower number of workers and batch size, interesting that he seems to get less interruptions by using a bigger batch, I'll try that next. 

I do get a slightly different traceback when I give it the `ctrl+c` though, but still related to the dataloader. 


```
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7f1a6a663910>                                                                                                                                  
Traceback (most recent call last):                                                                                                                                                                                         
  File "/home/$user/.local/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1466, in __del__                                                                                                           
    self._shutdown_workers()                                                                                                                                                                                               
  File "/home/$user/.local/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1430, in _shutdown_workers                                                                                                 
    w.join(timeout=_utils.MP_STATUS_CHECK_INTERVAL)                                                                                                                                                                        
  File "/home/$user/.conda/envs/ultralytics/lib/python3.10/multiprocessing/process.py", line 149, in join                                                                                                                
    res = self._popen.wait(timeout)                                                                                                                                                                                        
  File "/home/$user/.conda/envs/ultralytics/lib/python3.10/multiprocessing/popen_fork.py", line 40, in wait                                                                                                              
    if not wait([self.sentinel], timeout):                                                                                                                                                                                 
  File "/home/$user/.conda/envs/ultralytics/lib/python3.10/multiprocessing/connection.py", line 936, in wait                                                                                                             
    ready = selector.select(timeout)                                                                                                                                                                                       
  File "/home/$user/.conda/envs/ultralytics/lib/python3.10/selectors.py", line 416, in select                                                                                                                            
    fd_event_list = self._selector.poll(timeout)                                                                                                                                                                           
  File "/home/$user/.local/lib/python3.10/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler                                                                                                 
    _error_if_any_worker_fails()                                                                                                                                                                                           
RuntimeError: DataLoader worker (pid 395850) exited unexpectedly with exit code 2. Details are lost due to multiprocessing. Rerunning with num_workers=0 may give better error trace.
```

**Comment by wvalcke** [comment]
Bigger batch size -> means less iterations through the dataset, so less chance on a deadlock...

**Comment by CySlider** [comment]
I can confirm getting this bug with:

> Version: Ultralytics YOLOv8.0.57 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3090, 24257MiB)


```
Traceback (most recent call last):                                                                                             
  File "/home/tk/jupyter/panenv/bin/yolo", line 8, in <module>                                                                 
    sys.exit(entrypoint())                                                                                                     
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/yolo/cfg/__init__.py", line 318, in entrypoint        
    getattr(model, mode)(**overrides)  # default args from model                                                               
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/yolo/engine/model.py", line 327, in train             
    self.trainer.train()                                                                                                       
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/yolo/engine/trainer.py", line 186, in train           
    self._do_train(RANK, world_size)                                                                                           
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/yolo/engine/trainer.py", line 299, in _do_train       
    for i, batch in pbar:                                                                                                      
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/tqdm/std.py", line 1195, in __iter__                              
    for obj in iterable:                                                                                                       
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 628, in __next__            
    data = self._next_data()                                                                                                   
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1316, in _next_data         
    idx, data = self._get_data()                                                                                               
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1272, in _get_data          
    success, data = self._try_get_data()                                                                                       
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1120, in _try_get_data      
    data = self._data_queue.get(timeout=timeout)                                                                               
  File "/usr/lib/python3.10/queue.py", line 180, in get                                                                        
    self.not_empty.wait(remaining)                                                                                             
  File "/usr/lib/python3.10/threading.py", line 324, in wait                                                                   
    gotit = waiter.acquire(True, timeout)            
```                                                                          


It did even happen again at a different epoch (140 instead of 148) but same percentage (16%) when I started over. 
When I resumed training with last.pt, it did make it past this epoch.

I managed to train a similar model with 2000 epochs without issues.

> 
> Parametrs: yolo/engine/trainer: task=detect, mode=train, model=yolov8x.pt, data=data.yaml, epochs=3000, patience=200, batch=16, imgsz=(832, 480), save=True, save_period=100, cache=True, device=0, workers=8, project=./build, name=Test, exist_ok=True, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=True, cos_lr=False, close_mosaic=10, resume=False, amp=True, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.0, scale=0.1, shear=0.0, perspective=0.0, flipud=0.5, fliplr=0.5, mosaic=0.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=.

**Comment by lorenzomammana** [comment]
I get the same error when working with classification, it randomly hangs, it happened both when using cache and without using it.



`
yolo/engine/trainer: task=classify, mode=train, model=yolov8m-cls, data=path/to/data/, epochs=200, patience=20, batch=16, imgsz=640, save=True, save_period=-1, cache=True, device=0, workers=8, project=classification, name=, exist_ok=False, pretrained=True, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=classification/n_train
Overriding model.yaml nc=1000 with nc=6
`

```
  File "/home/lmammana/ultralytics/ultralytics/yolo/engine/model.py", line 329, in train
    self.trainer.train()
  File "/home/lmammana/ultralytics/ultralytics/yolo/engine/trainer.py", line 191, in train
    self._do_train(world_size)
  File "/home/lmammana/ultralytics/ultralytics/yolo/engine/trainer.py", line 360, in _do_train
    self.metrics, self.fitness = self.validate()
  File "/home/lmammana/ultralytics/ultralytics/yolo/engine/trainer.py", line 457, in validate
    metrics = self.validator(self)
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/lmammana/ultralytics/ultralytics/yolo/engine/validator.py", line 150, in __call__
    for batch_i, batch in enumerate(bar):
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/home/lmammana/ultralytics/ultralytics/yolo/data/build.py", line 39, in __iter__
    yield next(self.iterator)
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 628, in __next__
    data = self._next_data()
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1316, in _next_data
    idx, data = self._get_data()
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1272, in _get_data
    success, data = self._try_get_data()
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1120, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/queue.py", line 180, in get
    self.not_empty.wait(remaining)
  File "/home/lmammana/.conda/envs/aigord_1_4_0/lib/python3.9/threading.py", line 316, in wait
    gotit = waiter.acquire(True, timeout)
```

This is my env
```
Ultralytics YOLOv8.0.58 🚀 Python-3.9.15 torch-1.13.1+cu117 CUDA:0 (Tesla V100-DGXS-32GB, 32505MiB)
OS: Ubuntu 20.04
Python: 3.9
```

**Comment by hacktmz** [comment]
Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     80/400      29.9G     0.6966     0.3341      1.085         31        640: 100%|██████████| 351/351 [04:18<00:00,  1.36it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 45/45 [00:36<00:00,  1.23it/s]
                   all       4245       4769      0.965      0.961      0.986      0.838

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     81/400      29.9G     0.7018     0.3414      1.101        115        640:   3%|▎         | 9/351 [13:44:50<522:24:10, 5498.98s/it]
Traceback (most recent call last):
  File "/opt/conda/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/opt/conda/lib/python3.7/site-packages/ultralytics/yolo/cfg/__init__.py", line 318, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/opt/conda/lib/python3.7/site-packages/ultralytics/yolo/engine/model.py", line 329, in train
    self.trainer.train()
  File "/opt/conda/lib/python3.7/site-packages/ultralytics/yolo/engine/trainer.py", line 191, in train
    self._do_train(world_size)
  File "/opt/conda/lib/python3.7/site-packages/ultralytics/yolo/engine/trainer.py", line 302, in _do_train
    for i, batch in pbar:
  File "/opt/conda/lib/python3.7/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 681, in __next__
    data = self._next_data()
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1359, in _next_data
    idx, data = self._get_data()
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1315, in _get_data
    success, data = self._try_get_data()
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1163, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/opt/conda/lib/python3.7/queue.py", line 179, in get
    self.not_empty.wait(remaining)
  File "/opt/conda/lib/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7f5735ddf200>

**Comment by CySlider** [comment]
> Based on the error message you provided, it seems that the data loader is getting stuck when trying to fetch the next batch of data during validation. This could be caused by a few different factors, such as issues with the data itself or problems with the data loader configuration.
> 
> Here are a few things you can try to resolve the issue:
> 
>     1. Check your data: Make sure that the data you are using for validation is valid and contains no corrupt or missing files. You can also try running the training without validation for a few epochs to see if the error still occurs.
> 
>     2. Check your data loader configuration: Make sure that you have configured the data loader correctly and that the settings are appropriate for your dataset size and hardware. You can experiment with different values for `batch_size` and `num_workers` to see if that helps.
> 
>     3. Try disabling caching: If you are using caching, try disabling it to see if that resolves the issue. You can do this by setting the `cache` parameter to `False`.
> 
>     4. Check your system resources: Make sure you have enough system resources available to handle the training load. This includes CPU, GPU, and RAM. You can try reducing the batch size or the number of workers to lower the resource requirements.

1) Makes no sense as each epoch uses all data and this does not happen on first epoch. 
2) I gave my config paramters. I don't know where this could be wrong
3) I did try, didn't help
4) There are enough resources, and even if there were not, it should cause a different error.

> If none of these solutions work, you may want to try asking for help on the Ultralytics YOLOv8 GitHub Issues page, providing as much detail as possible about your setup and the error you are encountering.

Haha, has ChatGPT finally made it to issue responses :-)

**Comment by WeiChihChern** [comment]
still getting this.

**Comment by khanonenet** [comment]
I am also getting this issue at epoch 30 with the batch size of 8. Has anyone made any progress on this?

**Comment by harshdhamecha** [comment]
Any updates on this? For me, even Ctrl+C doesn't work

**Comment by glenn-jocher** [comment]
@harshdhamecha we understand that you are experiencing an issue with YOLOv8 during training where it gets stuck and even using Ctrl+C does not work. This could be due to various factors, such as issues with the data, configuration, or system resources. We suggest trying the following:

1. Verify your data: Ensure that your training data is valid and does not contain any corrupt or missing files. You can also try running the training without validation to see if the issue persists.

2. Check your data loader configuration: Make sure that you have correctly configured the data loader settings, such as `batch_size` and `num_workers`, based on your dataset size and hardware capabilities.

3. Disable caching: If you are using caching, try disabling it by setting the `cache` parameter to `False` and check if the problem persists.

4. Verify system resources: Ensure that you have enough system resources available, including CPU, GPU, and RAM, to handle the training load. You can try reducing the batch size or the number of workers to lower the resource requirements.

If you have tried these suggestions and are still facing the issue, we recommend posting your detailed setup and error encounter on the Ultralytics YOLOv8 GitHub Issues page to seek further assistance.

**Comment by vromero** [comment]
For whatever it's worth, I have the exact same error:

```
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     98/300      15.2G    0.09414    0.09607     0.7596         16        640:  82%|████████▏ | 58/71 [23:09<05:11, 23.96s/it]^C
Traceback (most recent call last):
  File "/usr/local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py", line 423, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py", line 376, in train
    self.trainer.train()
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py", line 192, in train
    self._do_train(world_size)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py", line 315, in _do_train
    for i, batch in pbar:
  File "/usr/local/lib/python3.10/dist-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/data/build.py", line 38, in __iter__
    yield next(self.iterator)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 633, in __next__
    data = self._next_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1328, in _next_data
    idx, data = self._get_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1284, in _get_data
    success, data = self._try_get_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1132, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/usr/lib/python3.10/queue.py", line 180, in get
    self.not_empty.wait(remaining)
  File "/usr/lib/python3.10/threading.py", line 324, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt
```
... on a Windows 11 with a 4090 when using (through a custom Docker image) cu118:

```
ARG BASE_IMAGE=docker.io/nvidia/cuda:12.2.0-devel-ubuntu22.04
ARG PYTHON_VERSION=3.8

FROM ${BASE_IMAGE}

RUN apt-get -o Acquire::Check-Valid-Until=false -o Acquire::Check-Date=false update
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        python3 \
        python3-pip \
        libgl1 \
        libglib2.0-0 libsm6 libxrender1 libxext6 
RUN rm -rf /var/lib/apt/lists/*


RUN pip3 install --extra-index-url https://download.pytorch.org/whl/cu118 \
        torch==2.0.1+cu118  \
        torchvision==0.15.2+cu118 \
        torchaudio

RUN pip install ultralytics
```

I can't ctrl-c but it goes beyond that because I can't even stop the docker container and it leaves my docker desktop in a rather unstable state:

![image](https://github.com/ultralytics/ultralytics/assets/1119553/ed8cf100-be07-441c-a7f3-edb16e403049)

**Comment by glenn-jocher** [comment]
@vromero from the information provided, it appears that your YOLOv8 training process is being interrupted during the data loading phase. This interruption is triggering a `KeyboardInterrupt` exception, which is what happens when a running Python program is stopped using a mechanism like Ctrl+C command or stopping a Docker container while the process is running.

While Ctrl+C is supposed to stop the process, it seems that it's not working as expected. This could be complicated further by the use of Docker, which adds another layer of complexity to the application's control flow.

A likely course of action could involve investigating your environment and around the data loader operation:

- **Data Integrity**: Please ensure the integrity of your data. Corrupted, missing, or otherwise malformed data files could potentially cause the loader to hang or fail. This includes checking the data paths and formats.

- **Dataloader Configuration**: Review your data loader parameters. Parameters like batch size or the number of workers can sometimes cause issues if not properly set based on the available resources.

- **Software and Driver compatibility**: Make sure that all your software components (CUDA, Pytorch, docker, etc.) are compatible with each other and your hardware. Incompatibility can cause inexpected errors and hangs.

- **Resources Usage**: Ensure that you have sufficient system resources (CPU, memory, storage etc.) and they are being properly allocated for the Docker environment to carry out the computation.

If the above suggestions cannot resolve the issue, I would suggest trying the training locally (not in Docker) to see if a similar issue arises. This might tell us whether the problem is related specifically to your Docker setup. If the problem still persists, you may need to provide more detail about the nature of your data (size, format, etc.) and your complete training and testing pipeline for further assistance.

**Comment by vromero** [comment]
Hi @glenn-jocher,

Thanks for your timely response for a comment in a closed issue and for the fantastic support you provide here in general.

The ctrl-c is me trying to actively close a training that was hanged. I'm almost sure its not a yolo thing given that docker itself become unreliable and I will surely try natively asap.

Thanks

**Comment by glenn-jocher** [comment]
Hi @vromero,

You're welcome, we're here to help!

Indeed, from the behaviour you've described, this seems more likely to be an issue with system resources being locked in an unstable state, rather than an issue with YOLOv8 itself. Running processes natively, instead of inside Docker containers, often gives more direct control on system resources, and could be helpful in debugging this issue.

When running training operations especially, it's not unusual for large amounts of system resources to be tied up. If these aren't released correctly for whatever reason (an unexpected exit, for example), it can result in the system or certain services becoming unresponsive.

Just as a precaution, after attempting the training natively, it's a good idea to monitor the system's resource usage to ensure that there are enough resources available for these training tasks. This can help in identifying if there're any resource related issues causing this hang.

Looking forward to hearing about your progress after trying these steps.

Best regards,

**Comment by UttamMahapatraToni** [comment]
I also faced this problem during trainning YOLOv8x model.  I also solved the problem
I think the problem is with the data loader configuration. To solve this problem , I reduced my batch_size from 16 to 4 . Then i could train the model with 100 epocs without any essue.

**Comment by glenn-jocher** [comment]
@UttamMahapatraToni glad to hear you've found a solution by adjusting the `batch_size` parameter. Resource constraints can sometimes lead to such issues, and reducing the `batch_size` typically alleviates memory pressure on the GPU. Should you face further challenges, don't hesitate to check back in. Happy training! 🚀

**Comment by mzahana** [comment]
I am also having this issue. I have RTX 4070 (12GB VRAM), plenty of RAM (128GB). 
Nvidia driver `535.154.05`
CUDA `12.2`
Pytorch version '2.1.2+cu121'
Ubuntu 20

 I have been using this setup for training over a month with no issues at all. Recently, I started to have this issue, sometimes it hangs at a specific epoch, sometimes it freezes my PC, and sometimes it breaks with segmentation fault error. I have no idea what is happening. I tried different batch sizes, and tried to disable the cache, but with no success.
 
 Any other suggestions to try ?
 Thanks

**Comment by glenn-jocher** [comment]
@mzahana hi there! It's quite puzzling that your previously stable setup is now experiencing these issues. 🤔 Since you've already tried adjusting batch sizes and disabling cache, let's consider a few more things:

1. **Software Updates**: Ensure all software dependencies are up-to-date, including PyTorch and CUDA. Sometimes, a mismatch after updates can cause instability.

2. **Hardware Check**: Run a hardware diagnostic to rule out any issues with your GPU or RAM.

3. **System Monitoring**: Keep an eye on system resource usage during training to spot any anomalies.

4. **Isolation Test**: Try running a minimal training script with a small dataset to see if the problem persists.

5. **Logs and Debugging**: Check system and application logs for any warnings or errors that might provide clues.

6. **Reinstallation**: As a last resort, consider reinstalling CUDA, PyTorch, and Nvidia drivers.

If the issue continues, feel free to share a snippet of your training command and any logs for a closer look. Hang in there, we'll get to the bottom of this! 💪

**Comment by balloch** [comment]
I can confirm I have this problem with a relatively small batch size (5gb of my 3090 full). Has anyone made any progress on this? I think it has something to do with Dataloader setup, but I am not sure

**Comment by glenn-jocher** [comment]
It might help to experiment with reducing the number of DataLoader workers or adjusting the `pin_memory` setting. These tweaks can sometimes resolve such issues.

**Comment by akshpv** [comment]
I experienced the same issue, and the reducing number of workers to just 1 resolved it

**Comment by glenn-jocher** [comment]
Good call—setting `workers=1` often sidesteps multiprocessing DataLoader hangs; you can set it like this (task-specific command shown) and see the workers arg in the Train settings docs: [Train settings (workers)](https://docs.ultralytics.com/modes/train/).

```bash
# Detection
yolo detect train data=data.yaml model=yolo11n.pt workers=1

# Classification
yolo classify train data=path/to/data/ model=yolo11n-cls.pt workers=1
```

```python
from ultralytics import YOLO
model = YOLO("your_model.pt")
model.train(data="data.yaml", workers=1)
```

If anyone still hits a hang, run once with `workers=0` to surface the true error, then open a new issue (this thread is closed) with `yolo checks` output and the full traceback for visibility.
